### PR TITLE
Transpile variables and UI refactor

### PR DESCRIPTION
**Pull request by KThompson-Lane**
## Changes:

1. Refactored GUI to improve user experience
2. Added parse tree visualisation within GUI
3. Added view for showing C# output
4. Added other various small animations/ QOL UI changes
5. Moved parser `NodeType` to enums folder
6. Made `ParseNode` fields public for accessibility (See comments for further discussion)
7. Added  `GetTerminals()` to `ParseNode` (As above)
8. Overrode `ParseNode` `ToString()`
9. Cleaned up some `TODO`'s that had already been completed
10. Added `TODO`'s 
11. Modified `Parser.cs` variable code to fix some errors. (Discussed below)
12. Added `Transpiler` support for variables with some caveats (as above)
13. Added simple unit test to `Transpiler`

### 6:
These fields could probably be replaced with properties, additionally it might be worth rethinking our constructors / adding additional helper methods for things like getting children at a certain index. Currently I manually do this in Transpiler, but having a method that throws an exception would aid in error handling.
### 7:
I've only used this in expression as I was lazy, it really shouldn't need to exist and I should do the tree traversal manually as I have with other types of node (see `TranspileDeclaration()`). I'm happy to remove it prior to merging or I can do that in a future PR
### 11:
I'll add these in as comments on individual lines for better visibility but I'll summarise here. Overall most of the variable parsing code worked great, just found a few hidden bugs that got merged. 
- The main issue was that the `VarAssignment()` method wasn't throwing exceptions when it ran into an incorrect token. It was logging errors however, but it needed to throw instead (/aswell? #19) 
- Additionally in `Statement()` the added try catches weren't returning after a successful parse which caused a few issues.
- A small change was just to use `VarIdentifier()` within `VarAssignment()` which removed duplicate code, additionally to match our BNF the check for a semicolon was moved to the end of `DeclareVariable()`

### 12:
There are 2 main caveats with variable support right now:
1. Due to our ignoring of whitespace, when the transpiler writes to the `Program.cs` file it doesn't include any spaces between tokens, resulting in `var i = 20;` becoming `vari=20;` which throws an error when compiled. I fixed this by just including a space between each token but this needs to be properly addressed within `Lexer` (perhaps with `Token.word` including preceding whitespaces?)

2. Our BNF/Parser currently allows users to declare variables without assigning them an initial value, this results in a compilation error within C# as by default C# requires implicit (`var`) variables to have an assignment to determine the type. We can either adapt our BNF to require the user to assign a value when first declaring a variable (easy method) or when the variable look-up table is implemented we can try to determine the type during the Transpilation stage based on the first value assigned to a variable. 

### Additional thoughts:
- Currently `Token` has a field for `Value` but this is pretty much redundant and used basically nowhere. It would be better to simply have "Word" and let the Transpiler deduce the value. Though this is sort of related to the above so happy to leave/discuss later. 

- Logging desperately needs a refactor see  #19

### CODE REVIEW

File: C_Flat_Interpreter/Common/ParseNode.cs [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-namespace C_Flat_Interpreter.Parser;
+namespace C_Flat_Interpreter.Common;
 
 public class ParseNode
 {
     private List<ParseNode> childNodes = new();
-    private Token? token;
-    private NodeType type;
+    public Token? token;
+    public NodeType type;
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Common/ParseNode.cs:9]: Might be worth changing [7-9] to properties?

File: C_Flat_Interpreter/Common/ParseNode.cs [hunk 2]
```diff
@@ -24,16 +24,28 @@ public void AddChild(ParseNode child)
...
-    
-    
+
     //Testing Function
     public List<ParseNode> getChildren()
     {
         return childNodes;
     }
 
+    public void GetTerminals(List<ParseNode> terminalList)
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Common/ParseNode.cs:34]: Not sure whether this is worth keeping, it mostly comes from my laziness and I probably need to do it differently in Transpiler anyways 😅

File: C_Flat_Interpreter/Parser/Parser.cs [hunk 3]
```diff
@@ -94,10 +94,10 @@ private bool Match(TokenType tokenType)
 		}
 		return true;
 	}
+	//TODO - Merge both together
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Parser/Parser.cs:97]: Was thinking here that Reset could be changed to do what Set does but with a default index of 0

File: C_Flat_Interpreter/Parser/Parser.cs [hunk 4]
```diff
@@ -145,37 +145,34 @@ public int Parse(List<Token> tokens)
 		}
 		
 		if (_currentIndex >= _totalTokens) return 0;
+		
+		//TODO - Improve logging overall
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Parser/Parser.cs:149]: #19

**Review by Spartan-Bailey — APPROVED** [review]

File: C_Flat_Interpreter/Transpiler/Transpiler.cs [hunk 5]
```diff
@@ -12,36 +12,186 @@ public class Transpiler : InterpreterLogger
...
+    private void PrintTerminal(ParseNode node)
+    {
+        var tokenToPrint = node.token ?? 
+                           throw new NullReferenceException("Node is marked as terminal but has no token");
+        if (tokenToPrint.Line > _currentLine)
+        {
+            _currentLine = tokenToPrint.Line;
+            Program += Environment.NewLine;
+        }
+        //Consider handling empty spaces within lexer
```
Comment (A-Russell23) [inline C_Flat_Interpreter/Transpiler/Transpiler.cs:33]: Is this something that I should be looking to work on soon?

**Review by A-Russell23 — APPROVED** [review]
Looks good. I've made a few notes but there isn't anything you need to act on right now. One thing you might want to consider at some point is ensuring that whatever is being displayed in the parse tree output window is cleared if you try to transpile something new that causes an error. Right now if I have the parse tree showing and then tranpile something invalid, the parse tree for the previous input still shows. Not a major issue but something worth changing to prevent any confusion

**Pull request merged**